### ISSUE TITLE
max_ids != max_contrib

### ISSUE THREAD

**Issue by mhauru** [opendp/smartnoise-sdk#544]
I'm setting one of my tables to have a non-default `max_ids`, and hitting the following error
```
  File "long-path-is-long\lib\site-packages\snsql\sql\private_reader.py", line 338, in _get_mechanisms
    assert(subquery.max_ids == max_contrib)

AssertionError
```
I'm unsure if I botched my metadata or if this a bug, but even if it's the former, the error message could be more informative.

I'm on `smartnoise-sql==0.2.10`.

**Comment by joshua-oss** [comment]
Hmm, that assert is meant to blow up if the rewritten query is not enforcing max_ids at the level required by the metadata.  It should not be possible for these to get out of sync in normal operation, since the metadata is attached to the query during the rewrite.  The assert was put in as a guardrail to catch any bugs or creative uses where the user might inadvertently rewrite with (e.g.) max_ids == 3, and then change this value to something smaller like 1 before adding noise.  Basically, the private_reader runs the subquery inside the database engine, and that is where reservoir sampling happens, then the private_reader adds noise outside the database, using a noise scale that must match the max_ids used in the reservoir sample.  If those don't match, it would be a privacy bug.

Can you comment on which of the two values is bigger when your assert fails?  We have unit tests that cover this and I'm not able to repro, so any additional context you could provide might help diagnose.

**Comment by mhauru** [comment]
Here's a case that does it for me. Metadata:
```
  schema_name:
    table_a:
      id:
        type: int
        private_id: True
      col_x:
        type: boolean
        private_id: False
    table_b:
      max_ids: 128
      id:
        type: int
        private_id: True
```
Query: ` SELECT count(*) FROM schema_name.table_a GROUP BY col_x`

In `PrivateReader._get_mechanisms`, `max_contrib` is 128 but `subquery.max_ids` is 1.

**Comment by joshua-oss** [comment]
Thanks for reporting this.  It should be fixed in smartnoise-sql==0.2.12.  Root cause was that the `private_reader` was initializing to use the largest `max_ids` in the metadata, before even knowing what table was going to be queried (because we only learn that at the time the caller calls `.execute()`.  This was old logic that was designed to allow callers to override the metadata `max_ids` but only revise downward, under the assumption that the data curator views `max_ids` as an upper bound on the number of ids that can be sampled.  We no longer allow callers to override this setting, so it was unnecessary and introduced bugs.  In no case could it have introduced a privacy bug (e.g. worst case is it would reservoir sample to a smaller number and then apply too much noise), but it was a pointless inconsistency that the assert thankfully caught.  Let me know if you still see issues.

**Comment by joshua-oss** [comment]
Fixed
